htlcswitch/link.go

```diff
@@ -1049,6 +1049,13 @@ func (l *channelLink) htlcManager() {
 						"data loss: %v", err)
 				}
 
+				// Before we fail the link, we'll send an error
+				// to the other party so they know we can't
+				// continue.
+				//
+				// TODO(roasbeef): use reliable sender here?
+				l.sendFatalError()
```

Comment:
I think this should be before the call to `MarkDataLoss`? I don't believe `MarkBorked` is called in peer/brontide.go for this since `PermanentFailure` will be false.  It also seems like an Error is already sent since `ErrRecoveryError` is used in the call to `l.fail(...)` below. Is the issue here that the p2p connection may go down _after_ `MarkDataLoss`, but _before_ we eventually call `ShouldSendToPeer` in `handleLinkFailure`?

Reply:
Yeah I actually had this initially, but the rationale was that by the time we send the `Error` they might have up already, which would cause some of the logic below to short circuit. 

> Is the issue here that the p2p connection may go down after MarkDataLoss, but before we eventually call ShouldSendToPeer in handleLinkFailure?

My understanding going into this change was that other implementations started to mostly ignore what we send for chan reest, even if it's all very invalid. So we now need to send the `Error` as that's the only thing they'll force close off of. 

IIUC, on restart, we'll also repeat this loop all over again, w/ the `MarkDataLoss` then being a noop.

Reply:
Tracing the pipeline a bit more, you're right that we'll eventually call `handleLinkFailure` in, which'll then fail since it's sending out `ErrRecoveryError`.